# PISA Data Exploration
## by Andreja Ho

## Preliminary Wrangling

> Briefly introduce your dataset here.

> This document explores a dataset containing PISA's 2012 data. PISA is a survey of students' skills and knowledge as they approach the end of compulsory education. It is not a conventional school test. Rather than examining how well students have learned the school curriculum, it looks at how well prepared they are for life beyond school. Around 510,000 students in 65 economies globally took part in the PISA 2012 assessment of reading, mathematics, and science representing about 28 million 15-year-olds globally.

In [5]:
# import all packages and set plots to be embedded inline
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sb

%matplotlib inline

In [6]:
# Read in original dataset and explore
df_origin = pd.read_csv('pisa2012.csv',sep=',', encoding='latin-1',error_bad_lines=False, index_col=False, dtype='unicode')
pd.set_option('display.max_columns', None)
df_origin.head(3)

Unnamed: 0      CNT SUBNATIO  STRATUM      OECD       NC SCHOOLID STIDSTD  \
0          1  Albania  0080000  ALB0006  Non-OECD  Albania  0000001   00001   
1          2  Albania  0080000  ALB0006  Non-OECD  Albania  0000001   00002   
2          3  Albania  0080000  ALB0006  Non-OECD  Albania  0000001   00003   

  ST01Q01 ST02Q01 ST03Q01 ST03Q02 ST04Q01                      ST05Q01  \
0      10       1      02    1996  Female                         No     
1      10       1      02    1996  Female  Yes, for more than one year   
2       9       1      09    1996  Female  Yes, for more than one year   

  ST06Q01    ST07Q01    ST07Q02    ST07Q03             ST08Q01 ST09Q01  \
0       6  No, never  No, never  No, never              None    None     
1       7  No, never  No, never  No, never  One or two times    None     
2       6  No, never  No, never  No, never              None    None     

  ST115Q01 ST11Q01 ST11Q02 ST11Q03 ST11Q04 ST11Q05 ST11Q06  \
0        1     Yes     Yes     Yes     Yes     NaN     NaN   
1        1     Yes     Yes     NaN     Yes     NaN     NaN   
2        1     Yes     Yes      No     Yes      No      No   

                 ST13Q01 ST14Q01 ST14Q02 ST14Q03 ST14Q04  \
0      <ISCED level 3A>       No      No      No      No   
1      <ISCED level 3A>      Yes     Yes      No      No   
2  <ISCED level 3B, 3C>      Yes     Yes     Yes      No   

                              ST15Q01            ST17Q01 ST18Q01 ST18Q02  \
0  Other (e.g. home duties, retired)   <ISCED level 3A>      NaN     NaN   
1        Working full-time <for pay>   <ISCED level 3A>       No      No   
2        Working full-time <for pay>   <ISCED level 3A>      Yes      No   

  ST18Q03 ST18Q04                       ST19Q01          ST20Q01  \
0     NaN     NaN   Working part-time <for pay>  Country of test   
1      No      No  Working full-time <for pay>   Country of test   
2     Yes     Yes  Working full-time <for pay>   Country of test   

           ST20Q02          ST20Q03 ST21Q01               ST25Q01 ST26Q01  \
0  Country of test  Country of test     NaN  Language of the test     Yes   
1  Country of test  Country of test     NaN  Language of the test     Yes   
2  Country of test  Country of test     NaN  Language of the test     Yes   

  ST26Q02 ST26Q03 ST26Q04 ST26Q05 ST26Q06 ST26Q07 ST26Q08 ST26Q09 ST26Q10  \
0      No     Yes      No      No      No      No     Yes      No     Yes   
1     Yes     Yes     Yes     Yes     Yes     Yes     Yes     Yes     Yes   
2     Yes     Yes     Yes      No     Yes     Yes     Yes     Yes     Yes   

  ST26Q11 ST26Q12 ST26Q13 ST26Q14  ST26Q15  ST26Q16  ST26Q17        ST27Q01  \
0      No     Yes      No     Yes  0008002  0008001  0008002            Two   
1     Yes     Yes     Yes     Yes  0008001  0008001  0008002  Three or more   
2      No     Yes      No     Yes  0008001  0008001  0008001  Three or more   

         ST27Q02        ST27Q03 ST27Q04 ST27Q05              ST28Q01  \
0            One           None    None    None          0-10 books    
1  Three or more  Three or more     Two     Two       201-500 books    
2            Two            Two     One     Two  More than 500 books   

    ST29Q01         ST29Q02   ST29Q03   ST29Q04         ST29Q05  \
0     Agree  Strongly agree     Agree     Agree           Agree   
1  Disagree  Strongly agree  Disagree  Disagree           Agree   
2     Agree  Strongly agree     Agree     Agree  Strongly agree   

          ST29Q06         ST29Q07         ST29Q08         ST35Q01  \
0           Agree           Agree  Strongly agree        Disagree   
1           Agree        Disagree        Disagree  Strongly agree   
2  Strongly agree  Strongly agree  Strongly agree  Strongly agree   

          ST35Q02   ST35Q03         ST35Q04         ST35Q05 ST35Q06  \
0           Agree  Disagree           Agree           Agree   Agree   
1  Strongly agree  Disagree           Agree        Disagree   Agree   
2  Strongly agree     Agree  Strongly ag

In [7]:
# Dataset shape
df_origin.shape

(485490, 636)

> Load in your dataset and describe its properties through the questions below.
Try and motivate your exploration goals through this section.
Here describe the whole dataset and steps what did you do to get a clean dataset.

PISA 2012 dataset is very complex and the original dataset contains 485490 rows and 636 columns. For this analysis, I’ve chosen only a few columns of interest. To understand the variables (column headers), I used PISA’s data dictionary[link] and PISA’s codebook[link]. Both documents were necessary to choose variables of interest and how to properly interpret them. 

First, I looked through the dictionary to learn what data was gathered and define the idea of the interest. Next, I read through the codebook to decide what features or variables are best suited for the analysis. When I choose the columns of interest I used Excel `textjoin()` function in order to get all columns’ names in a list, separated with a comma and quotation marks.
 
I am interested in how specific features impact students’ performance on math, reading, and science testing. The features if interests are:
    - wealth,<br> 
    - cultural and home possessions,<br>
    - educational resources<br>
    - sense of belongings<br>
    - parents’ and siblings’ presence<br>
    - the educational level of parents<br><br>
Besides those features, I will be analyzing if there is any difference between<br>
    - gender<br>
    - schools that joined OECD<br>
    - country<br><br>
Additionally, I would like to research if<br> 
    - mathematics anxiety<br>
is correlated with performance in math and science.



In [21]:
# Read in csv file:
# - select only columns of interest (col_list)
#           - to get all columns names in a list, separated with comma with quotation marks `textjoin()` function was used
# - issues with readind the large csv:
#           - this why code was used:pd.read_csv('pisa2012.csv',sep=',', encoding='latin-1',error_bad_lines=False, index_col=False, dtype='unicode')

col_list = ['CNT','OECD','SCHOOLID','ST04Q01','ST09Q01','ST115Q01','ST11Q01','ST11Q02','ST11Q03','ST11Q04','ANXMAT','CULTPOS','FISCED','HEDRES','HOMEPOS','MISCED','WEALTH','ANCBELONG','PV1MATH','PV1READ','PV1SCIE']
df = pd.read_csv('pisa2012.csv',sep=',', encoding='latin-1',error_bad_lines=False, index_col=False, dtype='unicode', usecols=col_list)

In [22]:
# Display all columns for better view and display first five columns
pd.set_option('display.max_columns', None)
df.head(5)

CNT      OECD SCHOOLID ST04Q01 ST09Q01 ST115Q01 ST11Q01 ST11Q02  \
0  Albania  Non-OECD  0000001  Female  None          1     Yes     Yes   
1  Albania  Non-OECD  0000001  Female  None          1     Yes     Yes   
2  Albania  Non-OECD  0000001  Female  None          1     Yes     Yes   
3  Albania  Non-OECD  0000001  Female  None          1     Yes     Yes   
4  Albania  Non-OECD  0000001  Female  None          2     Yes     Yes   

  ST11Q03 ST11Q04 ANXMAT CULTPOS             FISCED HEDRES HOMEPOS  \
0     Yes     Yes   0.32   -0.48  ISCED 3A, ISCED 4  -1.29   -2.61   
1     NaN     Yes    NaN    1.27  ISCED 3A, ISCED 4   1.12    1.41   
2      No     Yes    NaN    1.27        ISCED 5A, 6  -0.69    0.14   
3      No     Yes   0.31    1.27        ISCED 5A, 6   0.04   -0.73   
4     Yes     NaN   1.02    1.27  ISCED 3A, ISCED 4  -0.69   -0.57   

              MISCED WEALTH ANCBELONG   PV1MATH   PV1READ   PV1SCIE  
0  ISCED 3A, ISCED 4  -2.92   -0.7351  406.8469  249.5762  341.7009  
1        ISCED 5A, 6   0.69       NaN  486.1427  406.2936  548.9929  
2        ISCED 5A, 6  -0.23       NaN  533.2684    401.21  499.6643  
3        ISCED 3B, C  -1.17       NaN  412.2215   547.363  438.6796  
4               None  -1.17    0.8675  381.9209  311.7707  361.5628

In [23]:
df.shape

(485490, 21)

In [24]:
# Rename colum names for better understanding
df.rename(columns={'CNT':'Country_code', 'OECD':'OECD_country', 'SCHOOLID':'School_ID', 'ST04Q01':'Gender', 'ST09Q01':'Truancy_Skip_whole_school_day', 'ST115Q01':'Truancy_Skip_classes_within_school_day', 'ST11Q01':'At_Home_Mother', 'ST11Q02':'At_Home_Father', 'ST11Q03':'At_Home_Brothers', 'ST11Q04':'At_Home_Sisters', 'ANXMAT':'Mathematics_Anxiety', 'CULTPOS':'Cultural_Possessions', 'FISCED':'Educational_level_of_father_ISCED', 'HEDRES':'Home_educational_resources', 'HOMEPOS':'Home_Possessions', 'MISCED':'Educational_level_of_mother_ISCED', 'WEALTH':'Wealth', 'ANCBELONG':'Sense_of_Belonging_to_School_Anchored', 'PV1MATH':'Plausible_value_mathematics', 'PV1READ':'Plausible_value_reading', 'PV1SCIE':'Plausible_value_science'}, inplace=True)

In [25]:
# Check renamed columns
df.head(3)

Country_code OECD_country School_ID  Gender Truancy_Skip_whole_school_day  \
0      Albania     Non-OECD   0000001  Female                        None     
1      Albania     Non-OECD   0000001  Female                        None     
2      Albania     Non-OECD   0000001  Female                        None     

  Truancy_Skip_classes_within_school_day At_Home_Mother At_Home_Father  \
0                                      1            Yes            Yes   
1                                      1            Yes            Yes   
2                                      1            Yes            Yes   

  At_Home_Brothers At_Home_Sisters Mathematics_Anxiety Cultural_Possessions  \
0              Yes             Yes                0.32                -0.48   
1              NaN             Yes                 NaN                 1.27   
2               No             Yes                 NaN                 1.27   

  Educational_level_of_father_ISCED Home_educational_resources  \
0                 ISCED 3A, ISCED 4                      -1.29   
1                 ISCED 3A, ISCED 4                       1.12   
2                       ISCED 5A, 6                      -0.69   

  Home_Possessions Educational_level_of_mother_ISCED Wealth  \
0            -2.61                 ISCED 3A, ISCED 4  -2.92   
1             1.41                       ISCED 5A, 6   0.69   
2             0.14                       ISCED 5A, 6  -0.23   

  Sense_of_Belonging_to_School_Anchored Plausible_value_mathematics  \
0                               -0.7351                    406.8469   
1                                   NaN                    486.1427   
2                                   NaN                    533.2684   

  Plausible_value_reading Plausible_value_science  
0                249.5762                341.7009  
1                406.2936                548.9929  
2                  401.21                499.6643

Based on this [wikipedia](https://en.wikipedia.org/wiki/International_Standard_Classification_of_Education) site I have changed educational level of for better understanding:

| Pisa 2012 | From Wikipedia - ISCED 2011 | Equivalent Level - used in this dataset
| :- | :- | :-|
|ISCED 1           | Primary education | 1
|ISCED 2           | Lower secondary education|  2
|ISCED 3B, C       | Upper secondary education  | 3
|ISCED 3A, ISCED 4 | Post-secondary non-tertiary education   | 4
|ISCED 5B          | Short-cycle tertiary education | 5
|ISCED 5A, 6       | Bachelor's or equivalent  | 6

In [26]:
# Rename education level - father ()
df.Educational_level_of_father_ISCED.value_counts()

ISCED 3A, ISCED 4    118890
ISCED 5A, 6          113406
ISCED 2               66728
ISCED 5B              61617
ISCED 3B, C           39789
ISCED 1               35938
None                  16535
Name: Educational_level_of_father_ISCED, dtype: int64

In [27]:
# Rename education level - mother ()
df.Educational_level_of_mother_ISCED.value_counts()

ISCED 3A, ISCED 4    126768
ISCED 5A, 6          114452
ISCED 5B              68219
ISCED 2               66650
ISCED 1               36556
ISCED 3B, C           35672
None                  18768
Name: Educational_level_of_mother_ISCED, dtype: int64

In [37]:
# Replace educational values with numbers (see comments for explanation)
df["Educational_level_of_father_ISCED"] \
.replace({"ISCED 1": "1",
          "ISCED 2": "2",
          "ISCED 3B, C" : "3",
          "ISCED 3A, ISCED 4": "4",
          "ISCED 5B": "5",
          "ISCED 5A, 6" :"6"}, inplace=True)

In [38]:
# Replace educational values with numbers (see comments for explanation)
df["Educational_level_of_mother_ISCED"] \
.replace({"ISCED 1": "1",
          "ISCED 2": "2",
          "ISCED 3B, C" : "3",
          "ISCED 3A, ISCED 4": "4",
          "ISCED 5B": "5",
          "ISCED 5A, 6" :"6"}, inplace=True)

In [39]:
# Check for correct transformation 
df.Educational_level_of_mother_ISCED.value_counts()

4       126768
6       114452
5        68219
2        66650
1        36556
3        35672
None     18768
Name: Educational_level_of_mother_ISCED, dtype: int64

In [40]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 485490 entries, 0 to 485489
Data columns (total 21 columns):
 #   Column                                  Non-Null Count   Dtype 
---  ------                                  --------------   ----- 
 0   Country_code                            485490 non-null  object
 1   OECD_country                            485490 non-null  object
 2   School_ID                               485490 non-null  object
 3   Gender                                  485490 non-null  object
 4   Truancy_Skip_whole_school_day           479131 non-null  object
 5   Truancy_Skip_classes_within_school_day  479269 non-null  object
 6   At_Home_Mother                          460559 non-null  object
 7   At_Home_Father                          441036 non-null  object
 8   At_Home_Brothers                        400076 non-null  object
 9   At_Home_Sisters                         390768 non-null  object
 10  Mathematics_Anxiety                     314764 non-null 

In [43]:
pd.set_option('display.max_rows', None)
df.Country_code.value_counts().sort_index(ascending=False)

Vietnam                  4959
Uruguay                  5315
United Kingdom          12659
United Arab Emirates    11500
USA                     10294
Turkey                   4848
Tunisia                  4407
Thailand                 6606
Taiwan                   6046
Switzerland             11229
Sweden                   4736
Spain                   25313
Slovenia                 5911
Slovak Republic          4678
Singapore                5546
Serbia                   4684
Russia                   6992
Romania                  5074
Qatar                   10966
Portugal                 5722
Poland                   4607
Peru                     6035
Norway                   4686
New Zealand              4291
Netherlands              4460
Montenegro               4744
Mexico                  33806
Malaysia                 5197
Luxembourg               5258
Lithuania                4618
Liechtenstein             293
Latvia                   4306
Korea                    5033
Kazakhstan

In [41]:
# Combine names that have been broken down by states or cities with one unique name of the country
df = df.replace(['Connecticut (USA)', 'Florida (USA)', 'Massachusetts (USA)', 'United States of America'], 'USA')
df = df.replace(['Chinese Taipei'], 'Taiwan')
df = df.replace(['China-Shanghai', 'Hong Kong-China', 'Macao-China'], 'China')
df = df.replace(['Russian Federation', 'Perm(Russian Federation)'], 'Russia')

In [45]:
df.Country_code.value_counts().sort_index(ascending=False)

Vietnam                  4959
Uruguay                  5315
United Kingdom          12659
United Arab Emirates    11500
USA                     10294
Turkey                   4848
Tunisia                  4407
Thailand                 6606
Taiwan                   6046
Switzerland             11229
Sweden                   4736
Spain                   25313
Slovenia                 5911
Slovak Republic          4678
Singapore                5546
Serbia                   4684
Russia                   6992
Romania                  5074
Qatar                   10966
Portugal                 5722
Poland                   4607
Peru                     6035
Norway                   4686
New Zealand              4291
Netherlands              4460
Montenegro               4744
Mexico                  33806
Malaysia                 5197
Luxembourg               5258
Lithuania                4618
Liechtenstein             293
Latvia                   4306
Korea                    5033
Kazakhstan

### What is the structure of your dataset?

> Your answer here!

### What is/are the main feature(s) of interest in your dataset?

> Your answer here!

### What features in the dataset do you think will help support your investigation into your feature(s) of interest?

> Your answer here!

## Univariate Exploration

> In this section, investigate distributions of individual variables. If
you see unusual points or outliers, take a deeper look to clean things up
and prepare yourself to look at relationships between variables.

> Make sure that, after every plot or related series of plots, that you
include a Markdown cell with comments about what you observed, and what
you plan on investigating next.

### Discuss the distribution(s) of your variable(s) of interest. Were there any unusual points? Did you need to perform any transformations?

> Your answer here!

### Of the features you investigated, were there any unusual distributions? Did you perform any operations on the data to tidy, adjust, or change the form of the data? If so, why did you do this?

> Your answer here!

## Bivariate Exploration

> In this section, investigate relationships between pairs of variables in your
data. Make sure the variables that you cover here have been introduced in some
fashion in the previous section (univariate exploration).

### Talk about some of the relationships you observed in this part of the investigation. How did the feature(s) of interest vary with other features in the dataset?

> Your answer here!

### Did you observe any interesting relationships between the other features (not the main feature(s) of interest)?

> Your answer here!

## Multivariate Exploration

> Create plots of three or more variables to investigate your data even
further. Make sure that your investigations are justified, and follow from
your work in the previous sections.

### Talk about some of the relationships you observed in this part of the investigation. Were there features that strengthened each other in terms of looking at your feature(s) of interest?

> Your answer here!

### Were there any interesting or surprising interactions between features?

> Your answer here!

> At the end of your report, make sure that you export the notebook as an
html file from the `File > Download as... > HTML` menu. Make sure you keep
track of where the exported file goes, so you can put it in the same folder
as this notebook for project submission. Also, make sure you remove all of
the quote-formatted guide notes like this one before you finish your report!